### ISSUE TITLE
Explore: URLs with data source names lose PromQL

### ISSUE THREAD

**Issue by dgo-** [grafana/grafana#76752]
### What happened?

Explore URL that contains a data source name loses the PromQL query.

Example URL with data source name, which is not working: 
```
https://grafana.example.com/explore?left=%7B%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%2C%22queries%22%3A%5B%7B%22datasource%22%3A%7B%22type%22%3A%22prometheus%22%2C%22uid%22%3A%22test%22%7D%2C%22expr%22%3A%22up%7B%7D+%3D%3D+0%22%2C%22instant%22%3Afalse%2C%22range%22%3Atrue%2C%22refId%22%3A%22A%22%7D%5D%7D&orgId=5
```
<img width="452" alt="image" src="https://github.com/grafana/grafana/assets/1027969/d2b63935-7124-4842-a9ba-27c5d3aaa9d5">

The same URL with data source ID works:
```
https://grafana.example.com/explore?left=%7B%22range%22%3A%7B%22from%22%3A%22now-1h%22%2C%22to%22%3A%22now%22%7D%2C%22queries%22%3A%5B%7B%22datasource%22%3A%7B%22type%22%3A%22prometheus%22%2C%22uid%22%3A%22c91e9982-a5ec-4204-969c-b26ff2c597b7%22%7D%2C%22expr%22%3A%22up%7B%7D+%3D%3D+0%22%2C%22instant%22%3Afalse%2C%22range%22%3Atrue%2C%22refId%22%3A%22A%22%7D%5D%7D&orgId=5
```
<img width="452" alt="image" src="https://github.com/grafana/grafana/assets/1027969/083683ea-79b9-4bf8-af11-a2c6c18e6aef">

### What did you expect to happen?

I expected that the PromQL query from the URL is used in Grafana.

### Did this work before?

It worked until Grafana 10.0.2. 

### How do we reproduce it?

1. Use grafana explore URL with datasource name.

 

### Environment (with versions)?

Grafana:  10.1.5 
OS: Ubuntu 22
Browser: Firefox(118)/Chrome(118)
Mimir alertmanager (2.9.1)

As background we use the explore URL in our alert notification that are send out from the mimir alertmanager.
Our template looks like this:
```
    {{
        grafanaExploreURL "https://grafana.example.com"
        "test" "now-12h" "now" (queryFromGeneratorURL (index .Alerts 0).GeneratorURL)
    }
```

Currently we generate the data source names out of the mimir tenantID.  In the example I set the data source name static to "test". 

### Grafana platform?

A package manager (APT, YUM, BREW, etc.)

### Datasource(s)?

Mimir

**Comment by ifrost** [comment]
The first URL points to a data source with `uid:"test"` which based on the behavior probably doesn't exist on the instance. In case the data source with provided uid does not exist, Explore loads the default one. It may happen the default data source has the same type but we explicitly avoid keeping the query as it may result in an unexpected query against a different data source and may confuse users. 

What we should definitely do better is to inform the user what is happening and why. We have a task in our backlog to improve UX in such cases: https://github.com/grafana/grafana/issues/74687

**Comment by ifrost** [comment]
@grafana/alerting-squad IIRC you worked on a fix for it on your side. Did it make to 10.1.5?

**Comment by armandgrillet** [comment]
Someone in the alerting team fixed it for Mimir datasources but this is not related to Grafana Alerting: https://github.com/grafana/backend-enterprise/pull/5153

**Comment by dgo-** [comment]
Thanks for you feedback on the issue already.

Try to add some more informations:

If I add `uid: Mimir DMZ` Grafana will select the correct datasource, but drops the PromQL query.
<img width="815" alt="image" src="https://github.com/grafana/grafana/assets/1027969/68434761-835e-4804-bc05-17814aafa1af">

So if you give grafana the datasource name as uid. It will find the correct data source and select it but drop the PromQL query. 



Currently on Grafana 10.2.0.

**Comment by ifrost** [comment]
> If I add uid: Mimir DMZ Grafana will select the correct datasource, but drops the PromQL query.

Please make sure that uid is updated in both places - the root data source and the query, e.g.:

```
http://.../explore?&panes={"id":{"datasource":"gdev-prometheus","queries":[{"expr":"test",...,"datasource":{"uid":"gdev-prometheus"},"...&orgId=1
                                                ^                                                                     ^
                                                |---------------------------------- HERE AND HERE --------------------|
```

**Comment by dgo-** [comment]
Thanks for coming back to this. 

Our queries don't have a `panes` parameter.  The is also only a single data source. Here an example:
```
https://grafana/explore?left={"range":{"from":"now-12h","to":"now"},"queries":[{"datasource":{"type":"prometheus","uid":"Mimir"},"expr":"node_systemd_unit_state{name=\"service\",state=~\"failed|activating\"}+!=+0","instant":false,"range":true,"refId":"A"}]}&orgId=2
```

We create the URL with the mimir alertmanager functions `grafanaExploreURL` and `queryFromGeneratorURL`. Here a link to the [docs](https://grafana.com/docs/mimir/latest/references/architecture/components/alertmanager/#templating). 

Our template looks like this:
```
{{
  grafanaExploreURL "https://grafana/"
  "Mimir" "now-12h" "now" 
  (queryFromGeneratorURL (index .Alerts 0).GeneratorURL)
}}&orgId=2
```

**Comment by Elfo404** [comment]
I'm getting a bit lost, sorry. What is your datasource UID? and what is its name?

the generated URL should work fine even when there's no "root" datasource, ie.

```
?left={"queries":[{"refId":"A","expr":"dasd","range":true,"instant":true,"datasource":{"type":"prometheus","uid":"YOUR_DATSOURCE_UID"}}],"range":{"from":"now-1h","to":"now"}}
```

but note that i'm using the datsource UID in the `uid` field. using the datasource name may have worked in previous versions but wasn't the intended behaviour and not something we can properly do due to how the APIs we are using work.

https://github.com/grafana/grafana/assets/1170767/a76f15e0-e001-49e2-97f5-9bc0bed383da

forgot to record it on video, but also using grafana 10.2
<img width="208" alt="image" src="https://github.com/grafana/grafana/assets/1170767/39677671-38eb-4cea-a287-657fea151976">

in your original comment you mentioned that using the uid works as expected, so i don't see any issue here. maybe something for mimir docs to improve in https://grafana.com/docs/mimir/latest/references/architecture/components/alertmanager/#templating?

closing this for now as it seems everything works as intended, feel free to comment if i missed something and we can for sure reopen

**Comment by Alameli** [comment]
Thanks for explaining this, so is it possible to find out the exact version (or at least share the pr/commit which I prefer so I can see) to show when did it stop supporting using datasource names in the query? Asking because I have the same issue as the OP, where I was able to use the datasource name (instead of UID) and everything would be preserved in the explore queries.

Sorry for asking but I tried looking or the explanation/note in the release notes referring to the end of using datasource names to provide the explore links with the queries preserved.

**Comment by Elfo404** [comment]
this PR was when the use of names in the api we were using was originally deprecated: https://github.com/grafana/grafana/pull/33817

with https://github.com/grafana/grafana/pull/66286 we overhauled the URL <-> state sync in Explore (that before was simply  not working) and that's where using names stopped working if i'm not mistaken

**Comment by dgo-** [comment]
@Elfo404 Thanks for your reply. 

You are right that it is only an problem if you using the data source name as uid. I can understand that it is not expected that the datasource name will work in the uid field. 

In general it it really helpful feature to be able to use the datasource name in explore URL. Do you think it is an option to introduce a `name` field to select the datasource?
```
"datasource":{"type":"prometheus","name":"DATASOURCE_NAME"}
```
So you can use the datasource name or the datasource id.

In our case we are sending alerts with an explore link from the mimir alertmanager. The alertmanager is not aware of the data source IDs. For sure we could hardcode all IDs in the alertmanager template, but this will get a mess. Using the names is much more practical.

**Comment by Elfo404** [comment]
i don't think it's going to be possible, the reason the internal API is now using UIDs instead of names is (if i remember correctly) because af serious performance implications.

While i do understand your concerns, i would argue that what you are asking may become a pretty serious management headache. Datasource can change names over time - for any reason. UIDs are by definition way less (if at all) likely to change.
let's say you have an alert with an explore URL that uses a name, then someone in your org changes that datasource name, you are left with an alert that will open an explore link without any runnable query as there's no check in place to check if the name of the datasource was changed in grafana - possibly at the worst time.
This would not happen if using UIDs in your templates.

if you are provisioning grafana datasources you can even manually define such UIDs so that they can be more human readable (which is a recommended way of configuring grafana). Additionally i'm aware of some feature request about providing users with the ability to manually define UIDs in UI (at least when creating data sources, probably not when editing) ie. https://github.com/grafana/grafana/issues/82962 but i know this was brought up also internally.

In short, while I do really understand your concerns, this change is very unlikely to happen

**Comment by GroovyCarrot** [comment]
is there no chance of even just getting it to return a 302 Found and rewrite the url with the datasource uid?
